src/app/common/httpclient.ts

```diff
@@ -108,68 +95,72 @@ export class HttpClient {
         return options;
     }
 
-
-    public request(url: string, options?: RequestOptionsArgs, warn?: boolean): Promise<any> {
-        if (!this._conn) {
-            this._connectSvc.gotoConnect(true);
-            return Promise.reject("Not connected");
-        }
-
-        let req: Request;
-
+    private performRequest(conn: ApiConnection, url: string, options?: RequestOptionsArgs, warn?: boolean): Observable<Response> {
         let reqOpt = new RequestOptions(options);
 
         if (url.toString().indexOf("/") != 0) {
             url = '/' + url;
         }
 
-        reqOpt.url = this._conn.url + '/api' + url;
-        req = new Request(reqOpt);
-        
+        reqOpt.url = conn.url + '/api' + url;
+        let req = new Request(reqOpt);
         //
         // Set Access-Token
-        req.headers.set('Access-Token', 'Bearer ' + this._conn.accessToken);
-        
-        return this._http.request(req).toPromise()
-            .catch(e => {
+        req.headers.set('Access-Token', 'Bearer ' + conn.accessToken);
+        return this._http.request(req)
+            .catch((e, c) => {
                 // Status code 0 possible causes:
                 // Untrusted certificate
                 // Windows auth, prevents CORS headers from being accessed
                 // Service not responding
-                if (e.status == 0) {
-                    //
-                    // The first request to the API fails because windows auth has not started yet.
-                    // We repeat the request because in this case the next request will succeed.
-                    return this._http.request(req).toPromise()
-                        .catch(err => {
-                            //
-                            // Check to see if connected
-                            return this._http.options(this._conn.url).toPromise()
-                                .catch(e => {
-                                    this._connectSvc.reconnect();
-                                    return Promise.reject("Not connected");
+                if (e instanceof Response) {
+                    if (e.status == 0) {
+                        return this._http.request(req)
+                            .catch((err, _) => {
+                                // Check to see if connected
+                                return this._http.options(this._conn.url)
+                                    .catch((e, c2) => {
+                                        this._connectSvc.reconnect();
+                                        this.handleHttpError(err);
+                                        return c2
```

Comment:
When error happens, returning c2 seems not make sense. Please check if we can only throw error.